dodiscover/ci/kernel_test.py

```diff
@@ -0,0 +1,341 @@
+import numpy as np
+from scipy import stats
+from sklearn.metrics import pairwise_distances, pairwise_kernels
+from sklearn.metrics.pairwise import PAIRWISE_KERNEL_FUNCTIONS
+
+from .base import BaseConditionalIndependenceTest
+
+
+class KernelCITest(BaseConditionalIndependenceTest):
```

Comment:
Note that HSIC and KCI tests are also part of the GCM module: https://github.com/py-why/dowhy/blob/main/dowhy/gcm/independence_test/kernel.py

Reply:
Ooo nice! I was hoping to find time for the approx version. In line with the functional API push, would it be desired to refactor those into class objects with a required API to make all functionality that relies on CI testing more robust?

Where would the best place for "CI" tests to go?

Reply:
I am not sure if a class object is really necessary (at least, I don't see the benefit). In the GCM package, we ensure a compatible API by requiring the unconditional independence tests to be a function of the from ```Callable[[np.ndarray, np.ndarray], float]``` and for CI ```Callable[[np.ndarray, np.ndarray, np.ndarray], float]```. An example would be something like:

```
def hsic(X: np.ndarray, Y: np.ndarray, other_parameters...) -> float:
   .. compute p value...
  return p_value

def kci(X: np.ndarray, Y: np.ndarray, Z: np.ndarray, other_parameters...) -> float:
   .. compute p value...
  return p_value


def my_pc_algo(data: pd.DataFrame, independence_test : Callable[[np.ndarray, np.ndarray, np.ndarray], float] = hsic, conditional_test : Callable[[np.ndarray, np.ndarray, np.ndarray], float] = kci, significant_level: float = 0.05):
   ... do PC stuff ...
  if nodeC is None:
    p_value = independence_test(data[nodeA], data[nodeB])
  else:
    p_value = conditional_test(data[nodeA], data[nodeB], data[nodeC])

  reject = p_value < significant_level
  .. do PC stuff ...
```
If you want to configure the tests further, you can do simply:

```
my_pc_algo(data, partial(hsic, my_hsic_parameter_a = 0.1, kernel = 'laplace'))
```

Ok, just an idea to emphasize that we don't always need class objects for everything :)

Reply:
Btw. generally I like the idea to move the independence tests into a separate package. I think they can also be very helpful outside of the 'causality' context and are often required core functions in several algorithms (e.g. graph validation).

Reply:
Yeah agreed the CI testing is perhaps something to discuss next time then?

Reply:
**Re class vs no class:** I started out w/ functional too, but I realized I needed to pass in a large number of parameters to the structure learning algorithm. So it became really clunky. That was prolly my main motivation, but I agree (rn) that there's not much diff, except ease of reading and typing out. I'm down to switch back to function if that's desired.

E.g.

```Python
pc = PC(
    ... PC kwargs ...
    ci_test=kci,
    ... ci_test_kwargs
```

I think rn the PC algorithm as I have it has 7 parameters. If we also pass the ci test parameters, say for KCI, which has probably ~9 parameters. The PC algorithm overall has 16+ parameters possibly. I think eventually, it might be useful

Reply:
Yea, but we will have default parameters typically and then, in practice, only a fraction of users would really change more than 1 parameter overall :) In the 'average' use case, a user would not even change anything and sticks to the default ones. Btw. another way is also to create a config object that can be passed (just some ideas :)).

Reply:
Okay that makes sense. Ill refactor CI stuff then to just functions. And maybe eventually just a sep repo altogether.

Reply:
Are you ok w/ making that into a separate issue/PR?

Reply:
Yea sure. You can also keep the class structure for now and we can discuss this with the others first. I guess both (having it as class or as function) have their advantages and disadvantages. Maybe something for the next meeting?